adapters/tlsio_mbedtls.c

```diff
@@ -343,52 +348,86 @@ static int on_io_send(void *context, const unsigned char *buf, size_t sz)
+        mbedtls_entropy_free(&tls_io_instance->entropy);
+
+        tls_io_instance->tls_status = 0;
+    }
+}
+
+// Initialize mbedTLS
+static void mbedtls_init(TLS_IO_INSTANCE *tls_io_instance)
+{
+    if (tls_io_instance->tls_status == 1)
+    {
+        // Already initialized
+        return;
+    }
+    else if (tls_io_instance->tls_status == 2)
+    {
+        // The underlying connection has been closed, so here un-initialize first
+        mbedtls_uninit(tls_io_instance);
+    }
+
     const char *pers = "azure_iot_client";
 
     // mbedTLS initialize...
-    mbedtls_entropy_init(&instance->entropy);
-    mbedtls_ctr_drbg_init(&instance->ctr_drbg);
-    mbedtls_ssl_init(&instance->ssl);
-    mbedtls_ssl_session_init(&instance->ssn);
-    mbedtls_ssl_config_init(&instance->config);
-    mbedtls_x509_crt_init(&instance->trusted_certificates_parsed);
-    mbedtls_entropy_add_source(&instance->entropy, tlsio_entropy_poll, NULL, 128, 0);
-    mbedtls_ctr_drbg_seed(&instance->ctr_drbg, mbedtls_entropy_func, &instance->entropy, (const unsigned char *)pers, strlen(pers));
-    mbedtls_ssl_config_defaults(&instance->config, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT);
-    mbedtls_ssl_conf_rng(&instance->config, mbedtls_ctr_drbg_random, &instance->ctr_drbg);
-    mbedtls_ssl_conf_authmode(&instance->config, MBEDTLS_SSL_VERIFY_REQUIRED);
-    mbedtls_ssl_conf_min_version(&instance->config, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);          // v1.2
-    mbedtls_ssl_set_bio(&instance->ssl, instance, on_io_send, on_io_recv, NULL);
-    mbedtls_ssl_set_hostname(&instance->ssl, host);
-    mbedtls_ssl_set_session(&instance->ssl, &instance->ssn);
-
-    // DEPRECATED: debug functions do not belong in the tree.
-#if defined (MBED_TLS_DEBUG_ENABLE)
-    mbedtls_ssl_conf_dbg(&instance->config, mbedtls_debug, stdout);
+    mbedtls_x509_crt_init(&tls_io_instance->trusted_certificates_parsed);
+
+    mbedtls_entropy_init(&tls_io_instance->entropy);
+    mbedtls_entropy_add_source(&tls_io_instance->entropy, tlsio_entropy_poll, NULL, 128, 0);
+
+    mbedtls_ctr_drbg_init(&tls_io_instance->ctr_drbg);
+    mbedtls_ctr_drbg_seed(&tls_io_instance->ctr_drbg, mbedtls_entropy_func, &tls_io_instance->entropy, (const unsigned char *)pers, strlen(pers));
+
+    mbedtls_ssl_config_init(&tls_io_instance->config);
+    mbedtls_ssl_config_defaults(&tls_io_instance->config, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT);
+    mbedtls_ssl_conf_rng(&tls_io_instance->config, mbedtls_ctr_drbg_random, &tls_io_instance->ctr_drbg);
+    mbedtls_ssl_conf_authmode(&tls_io_instance->config, MBEDTLS_SSL_VERIFY_REQUIRED);
+    mbedtls_ssl_conf_min_version(&tls_io_instance->config, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3); // v1.2
+#if defined(MBED_TLS_DEBUG_ENABLE)
+    mbedtls_ssl_conf_dbg(&tls_io_instance->config, mbedtls_debug, stdout);
     mbedtls_debug_set_threshold(1);
 #endif
```

Comment:
Please remove this as per previous comment. We want to deprecate debug functions.